# statistics module

The statistics module provides a number of basic statistical functions including averages, variance and standard deviation.

The statistics module is normally imported using a 2 letter alias:

In [1]:
import statistics as st

The docstring of the statistics module is pretty detailed can be examined by inputting:

In [3]:
? st

Type:        module
String form: <module 'statistics' from 'c:\\Users\\pyip\\AppData\\Local\\mambaforge\\envs\\jupyterlab\\Lib\\statistics.py'>
File:        c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Docstring:  
Basic statistics module.

This module provides functions for calculating statistics of data, including
averages, variance, and standard deviation.

Calculating averages
--------------------

==================  ==================================================
Function            Description
==================  ==================================================
mean                Arithmetic mean (average) of data.
fmean               Fast, floating point arithmetic mean.
geometric_mean      Geometric mean of data.
harmonic_mean       Harmonic mean of data.
median              Median (middle value) of data.
median_low          Low median of data.
median_high         High median of data.
median_grouped      Median, or 50th percentile, of grouped d

As seen from the documentation, the module contains a collection of functions for calculating averages.

As seen from the documentation, the module contains a collection of functions for calculating averages:

|function|description|
|---|---|
|mean|Arithmetic mean (average) of data.|
|fmean|Fast, floating point arithmetic mean.|
|median|Median (middle value) of data.|
|median_low|Low median of data.|
|median_high|High median of data.|
|mode|Mode (most common value) of data.|
|multimode|List of modes (most common values of data).|
|quantiles|Divide data into intervals with equal probability.|
|geometric_mean|Geometric mean of data.|
|harmonic_mean|Harmonic mean of data.|
|median_grouped|Median, or 50th percentile, of grouped data.|

And for calculating the associated variability or spread of one dataset:

|function|description|
|---|---|
|pvariance|Population variance of data.|
|variance|Sample variance of data.|
|stdev|Sample standard deviation of data.|
|pstdev|Population standard deviation of data.|

Or of two datasets:

|function|description|
|---|---|
|covariance|Sample covariance for two variables.|
|correlation|Pearson's correlation coefficient for two variables.|
|linear_regression|Intercept and slope for simple linear regression.|

These statistical averages can be examined from a very simple dataset:

$$\text{data}=[\begin{matrix}7&5&3&1&2&4&6&4\end{matrix}]$$

In [5]:
data = [7, 5, 3, 1, 2, 4, 6, 4]

This dataset can be sorted:

$$\text{data}=\left[\begin{matrix}1&2&3&4&4&5&6&7\end{matrix}\right]$$

In [6]:
data.sort()

In [7]:
data

[1, 2, 3, 4, 4, 5, 6, 7]

The index of each element in the dataset can be shown. The length has a value of 1 higher than the last index. The last index is 7 so the length is 8:

$$\text{data}=\left[\begin{matrix}{\underset{0}{1}}&{\underset{1}{2}}&{\underset{2}{3}}&{\underset{3}{4}}&{\underset{4}{4}}&{\underset{5}{5}}&{\underset{6}{6}}&{\underset{7}{7}}\end{matrix}\right]$$

In [8]:
lendata = len(data)

In [9]:
lendata

8

The sum of the data can be calculated using:

$$\text{sum}=1+2+3+4+4+5+6+7=32$$

In [10]:
sumdata = sum(data)

In [11]:
sumdata

32

The integer mean can be calculated:

$$\text{intmean}=32//8=4$$

$$\text{intmod}=32﹪8=0$$

In [12]:
sum(data) // len(data)

4

In [13]:
sum(data) % len(data)

0

The floating point mean can be calculated:

$$\text{floatmean}=32/8=4.0$$

In [14]:
sum(data) / len(data)

4.0

The function mean returns the integer mean for no modulus and the float mean, when a modulus is present. The docstring can be examined using:

In [15]:
? st.mean

Signature:  st.mean(data)
Docstring:
Return the sample arithmetic mean of data.

>>> mean([1, 2, 3, 4, 4])
2.8

>>> from fractions import Fraction as F
>>> mean([F(3, 7), F(1, 21), F(5, 3), F(1, 3)])
Fraction(13, 21)

>>> from decimal import Decimal as D
>>> mean([D("0.5"), D("0.75"), D("0.625"), D("0.375")])
Decimal('0.5625')

If ``data`` is empty, StatisticsError will be raised.
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The mean of the dataset can be calculated using:

In [16]:
st.mean(data)

4

The function floating point mean, fmean on the other hand always returns the floating point mean. The docstring can be examined:

In [17]:
? st.fmean

Signature:  st.fmean(data, weights=None)
Docstring:
Convert data to floats and compute the arithmetic mean.

This runs faster than the mean() function and it always returns a float.
If the input dataset is empty, it raises a StatisticsError.

>>> fmean([3.5, 4.0, 5.25])
4.25
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The floating point mean of the dataset can be calculated using:

In [18]:
st.fmean(data)

4.0

The median is the middle point of the data which can be selected directly when the data is odd. When the data has an even number of values it is the mean of these two middle values:

$$\text{data}=\left[\begin{matrix}1&2&3&\textbf{4}&\textbf{4}&5&6&7\end{matrix}\right]$$

$$\text{median}=\frac{4+4}{2}=4$$

The docstring of the function median can be examined:

In [19]:
? st.median

Signature:  st.median(data)
Docstring:
Return the median (middle value) of numeric data.

When the number of data points is odd, return the middle data point.
When the number of data points is even, the median is interpolated by
taking the average of the two middle values:

>>> median([1, 3, 5])
3
>>> median([1, 3, 5, 7])
4.0
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The median of the dataset can be calculated using:

In [20]:
st.median(data)

4.0

Alternatively the value on the left is the median low value and can be read off directly:

$$\text{data}=\left[\begin{matrix}1&2&3&\textbf{4}&4&5&6&7\end{matrix}\right]$$

The docstring of the function median_low can be examined:

In [21]:
? st.median_low

Signature:  st.median_low(data)
Docstring:
Return the low median of numeric data.

When the number of data points is odd, the middle value is returned.
When it is even, the smaller of the two middle values is returned.

>>> median_low([1, 3, 5])
3
>>> median_low([1, 3, 5, 7])
3
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The median low of the dataset can be calculated using:

In [22]:
st.median_low(data)

4

And the value on the right is the median high value and can be read off directly:

$$\text{data}=\left[\begin{matrix}1&2&3&4&\textbf{4}&5&6&7\end{matrix}\right]$$

The docstring of the function median_high can be examined:

In [23]:
? st.median_high

Signature:  st.median_high(data)
Docstring:
Return the high median of data.

When the number of data points is odd, the middle value is returned.
When it is even, the larger of the two middle values is returned.

>>> median_high([1, 3, 5])
3
>>> median_high([1, 3, 5, 7])
5
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The median high of the dataset can be calculated using:

In [24]:
st.median_high(data)

4

The mode is the most commonly occuring value:

$$\text{data}=\left[\begin{matrix}1&2&3&\textbf{4}&\textbf{4}&5&6&7\end{matrix}\right]$$

The docstring of the function mode can be examined:

In [25]:
? st.mode

Signature:  st.mode(data)
Docstring:
Return the most common data point from discrete or nominal data.

``mode`` assumes discrete data, and returns a single value. This is the
standard treatment of the mode as commonly taught in schools:

    >>> mode([1, 1, 2, 3, 3, 3, 3, 4])
    3

This also works with nominal (non-numeric) data:

    >>> mode(["red", "blue", "blue", "red", "green", "red", "red"])
    'red'

If there are multiple modes with same frequency, return the first one
encountered:

    >>> mode(['red', 'red', 'green', 'blue', 'blue'])
    'red'

If *data* is empty, ``mode``, raises StatisticsError.
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The mode of the dataset can be calculated using:

In [26]:
st.mode(data)

4

The mode can be an important metric in a trial run. For example if a fashion designer wants to issue a prototype design of clothe, they may simplify manufacturing by starting only with the mode cloth size, as it has the largest customer base. If the trial of the first cloth batch is successful, they may complicate the manufacturer process to accomodate other sizes.

Sometimes a dataset will have multiple mode values, using mode will only give one of these values:

In [27]:
data2 = [1, 2, 2, 2, 2, 3, 4, 4, 4, 4, 6, 6, 7]

In [28]:
st.mode(data2)

2

There is an associated, multiple mode function. The docstring of the function multi_mode can be examined:

In [29]:
st.multimode(data2)

[2, 4]

The multiple mode function only retrieves modes that have the highest number of counts. It will not show modes that have only a slightly lower number of counts. For this purpose, recall that the Counter class is used:

In [30]:
from collections import Counter

In [31]:
Counter(data2).most_common()

[(2, 4), (4, 4), (6, 2), (1, 1), (3, 1), (7, 1)]

The concept of variance is to compute the average distance a datapoint differs from the mean.

Notice when when the sum of (each datapoint minus the mean) is calculated, the positive values and negative values cancel each other out and the result is 0:

$$\sum_{i=0}^{n}\left(x-\mu\right)=\left(1-4\right)+\left(2-4\right)+\left(3-4\right)+\left(4-4\right)+\left(4-4\right)+\left(5-4\right)+\left(6-4\right)+\left(7-4\right)$$

$$\left(-3\right)+\left(-2\right)+\left(-1\right)+\left(0\right)+\left(0\right)+\left(1\right)+\left(2\right)+(3)$$

$$-3-2-1+0+0+1+2+3$$

$$0$$

Instead the sum of (each datapoint minus the mean) squared is calculated. A negative value multiplied by itself is positive and this therefore always gives a positive value:

$$\sum_{i=0}^{n}\left(x-\mu\right)^2=\left(1-4\right)^2+\left(2-4\right)^2+\left(3-4\right)^2+\left(4-4\right)^2+\left(4-4\right)^2+\left(5-4\right)^2+\left(6-4\right)^2+\left(7-4\right)^2$$


$$\left(-3\right)^2+\left(-2\right)^2+\left(-1\right)^2+\left(0\right)^2+\left(0\right)^2+\left(1\right)^2+\left(2\right)^2+\left(3\right)^2$$

$$9+4+1+0+0+1+4+9$$

$$28$$

The mean is a normalised sum, and the population variance is the normalised sum of (each datapoint minus the mean) squared. The normalisation factor is the number of datapoints:

$$\frac{28}{8}=3.5$$

The variance takes into account a difference in the number of degrees of freedom and uses the number of datapoints minus one as a normalisation factor:

$$\frac{28}{(8-1)}=\frac{28}{7}=4.0$$

The docstring of the function pvariance and variance can be examined:

In [33]:
? st.pvariance

Signature:  st.pvariance(data, mu=None)
Docstring:
Return the population variance of ``data``.

data should be a sequence or iterable of Real-valued numbers, with at least one
value. The optional argument mu, if given, should be the mean of
the data. If it is missing or None, the mean is automatically calculated.

Use this function to calculate the variance from the entire population.
To estimate the variance from a sample, the ``variance`` function is
usually a better choice.

Examples:

>>> data = [0.0, 0.25, 0.25, 1.25, 1.5, 1.75, 2.75, 3.25]
>>> pvariance(data)
1.25

If you have already calculated the mean of the data, you can pass it as
the optional second argument to avoid recalculating it:

>>> mu = mean(data)
>>> pvariance(data, mu)
1.25

Decimals and Fractions are supported:

>>> from decimal import Decimal as D
>>> pvariance([D("27.5"), D("30.25"), D("30.25"), D("34.5"), D("41.75")])
Decimal('24.815')

>>> from fractions import Fraction as F
>>> pvariance([F(1, 4), F(5, 4), F

In [34]:
? st.variance

Signature:  st.variance(data, xbar=None)
Docstring:
Return the sample variance of data.

data should be an iterable of Real-valued numbers, with at least two
values. The optional argument xbar, if given, should be the mean of
the data. If it is missing or None, the mean is automatically calculated.

Use this function when your data is a sample from a population. To
calculate the variance from the entire population, see ``pvariance``.

Examples:

>>> data = [2.75, 1.75, 1.25, 0.25, 0.5, 1.25, 3.5]
>>> variance(data)
1.3720238095238095

If you have already calculated the mean of your data, you can pass it as
the optional second argument ``xbar`` to avoid recalculating it:

>>> m = mean(data)
>>> variance(data, m)
1.3720238095238095

This function does not check that ``xbar`` is actually the mean of
``data``. Giving arbitrary values for ``xbar`` may lead to invalid or
impossible results.

Decimals and Fractions are supported:

>>> from decimal import Decimal as D
>>> variance([D("27.5"), 

The population variance and variance can be calculated using:

In [35]:
st.pvariance(data)

3.5

In [36]:
st.variance(data)

4

Due to the squaring used when calculating the population variance or variance, these measurements have units of the mean squared. The standard deviation is the square root of the variance, which is measured in the same units as the mean:

$$\sqrt{3.5}=1.8708$$

The docstring of the function pstdev and stdev can be examined:

In [38]:
? st.pstdev

Signature:  st.pstdev(data, mu=None)
Docstring:
Return the square root of the population variance.

See ``pvariance`` for arguments and other details.

>>> pstdev([1.5, 2.5, 2.5, 2.75, 3.25, 4.75])
0.986893273527251
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

In [39]:
? st.stdev

Signature:  st.stdev(data, xbar=None)
Docstring:
Return the square root of the sample variance.

See ``variance`` for arguments and other details.

>>> stdev([1.5, 2.5, 2.5, 2.75, 3.25, 4.75])
1.0810874155219827
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

A distribution is normally quoted using the mean and standard deviation in the form:

$$4.0\pm2.0$$

The median value is less susceptible to large outliers and in such a scenario can be more accurate when describing a small dataset. For example if an outlier is purposely introduced, there is not much change in the median but the mean is highly distorted:

In [40]:
data.append(999)

In [41]:
data

[1, 2, 3, 4, 4, 5, 6, 7, 999]

In [42]:
st.mean(data)

114.55555555555556

In [43]:
st.median(data)

4

In [44]:
st.stdev(data)

331.67194300660674

The outlier can be removed from the data:

In [45]:
data.pop()

999

In [46]:
data

[1, 2, 3, 4, 4, 5, 6, 7]

The quantiles function is used to group data into quantiles, that is quarters with equal proability.

The docstring of the function quantiles can be examined:

In [47]:
? st.quantiles

Signature:  st.quantiles(data, *, n=4, method='exclusive')
Docstring:
Divide *data* into *n* continuous intervals with equal probability.

Returns a list of (n - 1) cut points separating the intervals.

Set *n* to 4 for quartiles (the default).  Set *n* to 10 for deciles.
Set *n* to 100 for percentiles which gives the 99 cuts points that
separate *data* in to 100 equal sized groups.

The *data* can be any iterable containing sample.
The cut points are linearly interpolated between data points.

If *method* is set to *inclusive*, *data* is treated as population
data.  The minimum value is treated as the 0th percentile and the
maximum value is treated as the 100th percentile.
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The quantile values can be calculated by:

In [48]:
data3 = [1, 1, 2, 2, 3, 4, 4, 4, 5, 6, 7, 8, 8, 8, 9, 9]

In [49]:
data3

[1, 1, 2, 2, 3, 4, 4, 4, 5, 6, 7, 8, 8, 8, 9, 9]

In [51]:
st.quantiles(data3)

[2.25, 4.5, 8.0]

The quantile values can be used to cut the data into quantiles using:

In [52]:
q1 = []
q2 = []
q3 = []
q4 = []
for num in data3:
    if(num < st.quantiles(data3)[0]):
        q1.append(num)
    elif (num < st.quantiles(data3)[1]):
        q2.append(num)
    elif (num < st.quantiles(data3)[2]):
        q3.append(num)
    else:
        q4.append(num)
print(q1, q2, q3, q4, sep='\n')

[1, 1, 2, 2]
[3, 4, 4, 4]
[5, 6, 7]
[8, 8, 8, 9, 9]


The mean is calculated using a normalised sum. The geometric mean instead uses a normalised product:

$$\text{data}=\left[\begin{matrix}1&2&3&4&4&5&6&7\end{matrix}\right]$$

$$\text{prod}=1\ast2\ast3\ast4\ast4\ast5\ast6\ast7=20160$$

$$\text{len}=8$$

To normalise the product the nth root is taken:

$$\left(1\ast2\ast3\ast4\ast4\ast5\ast6\ast7\right)\ast\ast(1/8)=3.4519$$

The docstring of the function geometric mean can be examined:

In [53]:
? st.geometric_mean

Signature:  st.geometric_mean(data)
Docstring:
Convert data to floats and compute the geometric mean.

Raises a StatisticsError if the input dataset is empty,
if it contains a zero, or if it contains a negative value.

No special efforts are made to achieve exact results.
(However, this may change in the future.)

>>> round(geometric_mean([54, 24, 36]), 9)
36.0
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The geometric mean can be calculated using:

In [54]:
st.geometric_mean(data)

3.451924719784297

For this dataset it can be manually calculated using:

In [55]:
(1*2*3*4*4*5*6*7) ** (1/8)

3.451924719784297

The harmonic mean is calculated using the inverse sum. 

$$\text{data}=\left[\begin{matrix}1&2&3&4&4&5&6&7\end{matrix}\right]$$


$$\text{invsum}=\left(\frac{1}{1}+\frac{1}{2}+\frac{1}{3}+\frac{1}{4}+\frac{1}{4}+\frac{1}{5}+\frac{1}{6}+\frac{1}{7}\right)$$

This is normalised by the length:

$$\text{len}=8$$

$$\frac{\left(\frac{1}{1}+\frac{1}{2}+\frac{1}{3}+\frac{1}{4}+\frac{1}{4}+\frac{1}{5}+\frac{1}{6}+\frac{1}{7}\right)}{8}$$

Then finally the inverse of this is taken:

$$\frac{8}{\left(\frac{1}{1}+\frac{1}{2}+\frac{1}{3}+\frac{1}{4}+\frac{1}{4}+\frac{1}{5}+\frac{1}{6}+\frac{1}{7}\right)}$$

The docstring of the function harmonic mean can be examined:

In [56]:
? st.harmonic_mean

Signature:  st.harmonic_mean(data, weights=None)
Docstring:
Return the harmonic mean of data.

The harmonic mean is the reciprocal of the arithmetic mean of the
reciprocals of the data.  It can be used for averaging ratios or
rates, for example speeds.

Suppose a car travels 40 km/hr for 5 km and then speeds-up to
60 km/hr for another 5 km. What is the average speed?

    >>> harmonic_mean([40, 60])
    48.0

Suppose a car travels 40 km/hr for 5 km, and when traffic clears,
speeds-up to 60 km/hr for the remaining 30 km of the journey. What
is the average speed?

    >>> harmonic_mean([40, 60], weights=[5, 30])
    56.0

If ``data`` is empty, or any element is less than zero,
``harmonic_mean`` will raise ``StatisticsError``.
File:      c:\users\pyip\appdata\local\mambaforge\envs\jupyterlab\lib\statistics.py
Type:      function

The harmonic mean can be calculated using:

In [57]:
st.harmonic_mean(data)

2.814070351758794

For this dataset it can be manually calculated using:

In [58]:
8 / ((1/1)+(1/2)+(1/3)+(1/4)+(1/4)+(1/5)+(1/6)+(1/7))

2.814070351758794

Supposing the following grouped data is obtained from a histogram of values:

$$\text{data4} = \left[\begin{matrix}2&2&3&3&3&4\end{matrix}\right]$$


Then for each individual datapoint the range is between:

|central value|range|
|---|---|
|2|$1.5\text{ to } 2.5$|
|2|$1.5\text{ to } 2.5$|
|3|$2.5\text{ to } 3.5$|
|3|$2.5\text{ to } 3.5$|
|3|$2.5\text{ to } 3.5$|
|4|$3.5\text{ to } 4.5$|

There are 6 datapoints so $n=6$ and each datapoint spans an interval $i=1$. 

The median range is $2.5\text{ to } 3.5$, this has a lower limit of $l=2.5$ and has a frequency $f=3$.

There are 2 datapoints before the median range, so $b=2$.

The median grouped is defined as:

$$l+\frac{\left(\frac{n}{2}-b\right)}{f}\ast i$$

$$2.5+\frac{\left(\frac{6}{2}-2\right)}{3}\ast1$$

$$2.833$$

The docstring of the function median_grouped can be examined:

In [60]:
? st.median_grouped

Signature:  st.median_grouped(data, interval=1.0)
Docstring:
Estimates the median for numeric data binned around the midpoints
of consecutive, fixed-width intervals.

The *data* can be any iterable of numeric data with each value being
exactly the midpoint of a bin.  At least one value must be present.

The *interval* is width of each bin.

For example, demographic information may have been summarized into
consecutive ten-year age groups with each group being represented
by the 5-year midpoints of the intervals:

    >>> demographics = Counter({
    ...    25: 172,   # 20 to 30 years old
    ...    35: 484,   # 30 to 40 years old
    ...    45: 387,   # 40 to 50 years old
    ...    55:  22,   # 50 to 60 years old
    ...    65:   6,   # 60 to 70 years old
    ... })

The 50th percentile (median) is the 536th person out of the 1071
member cohort.  That person is in the 30 to 40 year old age group.

The regular median() function would assume that everyone in the
tricenarian age group wa

The only keyword argument is the interval which has a default value of 1.0. All the rest of the parameters can be inferred from the data. The median grouped of the data can be calculated using:

In [61]:
data4 = [2, 2, 3, 3, 3, 4]

In [62]:
st.median_grouped(data4, interval=1)

2.8333333333333335

In [63]:
2.5 + (((6 / 2) - 2) / 3) * 1

2.8333333333333335

The covariance is a measure of the joint variability of two datasets of the same size. To examine this the following lists will be created and their respective mean and variance will be calculated: 

In [64]:
data5 = [1, 2, 3, 4]

In [65]:
mean_data5 = st.mean(data5)

In [67]:
mean_data5

2.5

In [66]:
var_data5 = st.variance(data5)

In [68]:
var_data5

1.6666666666666667

In [69]:
data6 = [1, 2, 3, 2]

In [73]:
mean_data6 = st.mean(data6)

In [74]:
mean_data6

2

In [75]:
var_data6 = st.variance(data6)

In [76]:
var_data6

0.6666666666666666

The covariance of two datasets is similar to the variance of a single dataset. Instead of summing the squared differences from the mean, of a single dataset. The sum of the product of the difference from the mean of the first dataset is multiplied by the difference from the mean of the second dataset:

$$\sum_{i=0}^{n}{\left(x_i-\mu_x\right)\ast(y_i-\mu_y)}=\left(1-2.5\right)\ast\left(1-2\right)+\left(2-2.5\right)\ast\left(2-2\right)+\left(3-2.5\right)\ast\left(3-2\right)+\left(4-2.5\right)\ast(2-2)=2.0$$

The covariance takes into account a difference in the number of degrees of freedom and uses the number of datapoints minus one as a normalisation factor:

$$\frac{2.0}{4-1}=\frac{2}{3}=0.667$$

The docstring of the function covariance can be examined: